File: 2006-find-the-student-that-will-replace-the-chalk/2006-find-the-student-that-will-replace-the-chalk.java
class Solution {
    public int chalkReplacer(int[] chalk, int k) {
        int n = chalk.length;
        long sum = 0;
        long temp[] = new long[n];
        for(int i = 0; i < n; i++){
            sum += chalk[i];
            temp[i] += sum;
        }

        long tar = k % sum;
        int low = 0, high = n - 1, ans = 0;
        while(low <= high){
            int mid = low + (high - low) / 2;
            if(temp[mid] == tar){
                return mid + 1;
            } else if(temp[mid] < tar){
                low = mid + 1;
            } else{
                ans = mid;
                high = mid - 1;
            }
        }

        return ans;
    }
}

File: 2006-find-the-student-that-will-replace-the-chalk/2006-find-the-student-that-will-replace-the-chalk.cpp
class Solution {
public:
    int chalkReplacer(vector<int>& chalk, int k) {
        vector<long long> temp;;
        long long sum = 0;
        for(int i : chalk){
            sum += i;
            temp.push_back(sum);
        }

        k %= sum;
        int l = 0, h = chalk.size() - 1, ans = 0;
        while(l <= h){
            int m = l + (h - l) / 2;
            if(temp[m] == k){
                return m + 1;
            } else if(temp[m] < k){
                l = m + 1;
            } else{
                ans = m;
                h = m - 1;
            }
        }

        return ans;
    }
};